File: Restaurant/JHTSReview.cpp
//
//  JHTSReview.cpp
//  Restaurant
//
//  Created by Jonathan Hoyle on 12/21/17.
//  Copyright © 2017 Jonathan Hoyle. All rights reserved.
//

#include "JHTSReview.h"
#include "JHTSRestaurant.h"
#include "JHTSUser.h"


// Constructor
JHTSReview::JHTSReview(int starRating, string reviewText, JHTSGUID userGUID, JHTSGUID restaurantGUID)
{
    mStarRating = starRating;
    mReviewText = reviewText;
    mUserGUID = userGUID;
    mRestaurantGUID = restaurantGUID;
    mMostRecentUpdate = time(NULL);
}

// Virtual Destructor
JHTSReview::~JHTSReview()
{
}

// Reconstructor
void JHTSReview::operator()(int starRating, string reviewText)
{
    mStarRating = starRating;
    mReviewText = reviewText;
    mMostRecentUpdate = time(NULL);
}

void JHTSReview::setStarRating(int starRating)
{
    mStarRating = starRating;
    mMostRecentUpdate = time(NULL);
}

void JHTSReview::setReviewText(const string &reviewText)
{
    mReviewText = reviewText;
    mMostRecentUpdate = time(NULL);
}

// Get a Review from the database for user and restaurant
JHTSReview *JHTSReview::getReview(JHTSGUID userGUID, JHTSGUID restaurantGUID)
{
    // Look for it in our list
    for (auto ptr = mDatabaseObjectList.begin(); ptr != mDatabaseObjectList.end(); ptr++)
    {
        JHTSReview *review = dynamic_cast<JHTSReview *>(*ptr);
        
        if ((review != NULL) && (review->mUserGUID == userGUID) && (review->mRestaurantGUID == restaurantGUID))
            return review;
    }
    
    return NULL;
}

bool JHTSReview::isValid() const
{
    if (!JHTSDatabaseObject::isValid())
        return false;
    
    JHTSUser *user = JHTSUser::getUser(mUserGUID);
    if (user == NULL) return false;
    set<JHTSGUID> userReviewList = user->listOfReviews();
    if (userReviewList.find(mGUID) == userReviewList.end())
        return false;
    
    JHTSRestaurant *restaurant = JHTSRestaurant::getRestaurant(mRestaurantGUID);
    if (restaurant == NULL) return false;
    set<JHTSGUID> restaurantReviewList = restaurant->listOfReviews();
    if (restaurantReviewList.find(mGUID) == restaurantReviewList.end())
        return false;

    return true;
}

void JHTSReview::print(ostream &stream) const
{
    stream << "Rating: " << mStarRating << endl;
    stream << "Text: \"" << mReviewText << "\"" << endl;
    
    JHTSUser *user = JHTSUser::getUser(mUserGUID);
    stream << "by: " << *user << endl;
    
    JHTSRestaurant *restaurant = JHTSRestaurant::getRestaurant(mRestaurantGUID);
    stream << "for: " << *restaurant << endl;
}

File: JHTSReview.cs
using System;
using System.Collections.Generic;
using RestaurantReviews.Object;
using RestaurantReviews.Restaurant;
using RestaurantReviews.User;

namespace RestaurantReviews.Review
{
    public class JHTSReview : JHTSObject
    {
        // Properties
        public int UserGUID { get; private set; }
        public int RestaurantGUID { get; private set; }
        public DateTime MostRecentUpdate { get; private set; }

        private int mStarRating;
        public int StarRating
        {
            get => mStarRating;
            set { mStarRating = value; MostRecentUpdate = DateTime.Now; }
        }

        private string mReviewText;
        public string ReviewText
        {
            get => mReviewText;
            set { mReviewText = value; MostRecentUpdate = DateTime.Now; }
        }

        // Internal Constructor
        protected internal JHTSReview(int starRating, string reviewText, int userGUID, int restaurantGUID)
        {
            mStarRating = starRating;
            mReviewText = reviewText;
            UserGUID = userGUID;
            RestaurantGUID = restaurantGUID;
            MostRecentUpdate = DateTime.Now;

            if (mCachedListOfReviews != null)
                mCachedListOfReviews.Add(GUID);

            if (!IsValid)
            {
                clearCache();
                updateCache();
            }
        }

        // Destructor
        ~JHTSReview()
        {
            if (mCachedListOfReviews != null)
                mCachedListOfReviews.Remove(GUID);
        }

        // List of all review objects
        //  (lazy loaded, don't calculate until needed)
        static private SortedSet<int> mCachedListOfReviews = null;
        static public SortedSet<int> ListOfReviews
        {
            get
            {
                if (mCachedListOfReviews == null)
                    mCachedListOfReviews = listByType<JHTSReview>();

                return mCachedListOfReviews;
            }
        }

        // Get a review by guid
        static public JHTSReview getReview(int guid)
        { return (objectAtGUID(guid) as JHTSReview); }

        // Get a review by user/restaurant combination
        static public JHTSReview getReview(int userGUID, int restaurantGUID)
        {
            foreach (var reviewGUID in ListOfReviews)
            {
                JHTSReview review = getReview(reviewGUID);

                if ((review != null) && (review.UserGUID == userGUID) && (review.RestaurantGUID == restaurantGUID))
                    return review;
            }

            return null;
        }

        // Clear the object's cached data
        protected override void clearCache()
        {
            base.clearCache();

            if (mCachedListOfReviews != null)
            {
                mCachedListOfReviews.Clear();
                mCachedListOfReviews = null;
            }
        }

        // Update the object's cached data
        protected override void updateCache()
        {
            base.updateCache();
        }

       // Validity check
        protected override bool IsValid
        {
            get
            {
                if (!base.IsValid)
                    return false;

                // Verify that the associated user exists
                JHTSUser user = JHTSUser.getUser(UserGUID);
                if (user == null)
                    return false;

                // Verify that the associated user has this review
                SortedSet<int> userList = user.ListOfReviews;
                if (!userList.Contains(GUID))
                    return false;

                // Verify that the associated restaurant exists
                JHTSRestaurant restaurant = JHTSRestaurant.getRestaurant(RestaurantGUID);
                if (restaurant == null)
                    return false;

                // Verify that the associated restaurant has this review
                SortedSet<int> restaurantList = restaurant.ListOfReviews;
                if (!restaurantList.Contains(GUID))
                    return false;

                return true;
            }
        }

        // Writing out contents
        public override string ToString()
        {
            string output = base.ToString() + Environment.NewLine;

            output += "Rating: " + StarRating + Environment.NewLine;
            output += "Text: \"" + ReviewText + "\"" + Environment.NewLine;
            output += "by User #" + UserGUID + Environment.NewLine;
            output += "for Restaurant #" + RestaurantGUID + Environment.NewLine;

            return output;
        }
    }
}